### PR TITLE
tunnel: add support for websocket connections

### PR DESCRIPTION
**Pull request by rjeczalik**
~~deps: https://github.com/koding/tunnel/pull/19~~

/cc @cihangir

### CODE REVIEW

**Comment by rjeczalik** [comment]
ping @cihangir 

race is going to be fixed with #19, when merged I'll rebase this one :boom::boom:

File: server.go [hunk 1]
```diff
@@ -582,6 +615,48 @@ func (s *Server) checkConnect(fn func(w http.ResponseWriter, r *http.Request) er
...
+func isWebsocketConn(r *http.Request) bool {
+	return r.Method == "GET" && headerContains(r.Header["Connection"], "upgrade") &&
+		headerContains(r.Header["Upgrade"], "websocket")
+}
+
+// headerContains is a copy of tokenListContainsValue from gorilla/websocket/util.go
+func headerContains(header []string, value string) bool {
+	for _, h := range header {
+		for _, v := range strings.Split(h, ",") {
+			if strings.EqualFold(strings.TrimSpace(v), value) {
```
Comment (cihangir) [inline server.go]: excellent! -> EqualFold

Comment (cihangir) [inline server.go]: ops didnt see your comment :D  https://github.com/koding/tunnel/pull/14/files#diff-34c6b408d72845d076d47126c29948d1R637

File: helper_test.go [hunk 2]
```diff
@@ -185,10 +209,72 @@ func echoTCPIdent(tt *tunneltest.TunnelTest, echo, ident string) (string, error)
...
+
+func handlerLatencyEchoWS(t *testing.T) func(w http.ResponseWriter, r *http.Request) {
+	return handlerEchoSleepWS(t, true)
+}
+
+func handlerEchoSleepWS(t *testing.T, doSleep bool) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (cihangir) [inline helper_test.go]: in most of the cases i dont use both http error responses and test failures at the same time. returning the error as handler response and checking the return value in the calling function seems cleaner to me.

File: helper_test.go [hunk 3]
```diff
@@ -185,10 +209,72 @@ func echoTCPIdent(tt *tunneltest.TunnelTest, echo, ident string) (string, error)
...
+				continue
+			}
+
+			if doSleep {
+				sleep()
+			}
+
+			err = conn.WriteJSON(&msg)
+			if err != nil {
+				t.Errorf("WriteJSON error: %s", err)
```
Comment (cihangir) [inline helper_test.go]: would be better to close the connection in the case of errors, ( setting `msg.Close` to true )

**Comment by cihangir** [comment]
You can rebase this one now @rjeczalik

File: helper_test.go (hunk 2 above)
Comment (rjeczalik) [inline helper_test.go]: done

File: helper_test.go (hunk 3 above)
Comment (rjeczalik) [inline helper_test.go]: done

**Comment by rjeczalik** [comment]
@cihangir rebased, ptal

**Comment by cihangir** [comment]
:+1:

**Pull request merged**